File: understand-anything-plugin/packages/dashboard/src/core/embedding-search.ts
import type { GraphNode } from "./types.js";
import type { SearchResult, SearchOptions } from "./search.js";

/**
 * Options for a semantic search. Extends the lexical {@link SearchOptions}
 * with a similarity `threshold` so callers can drop weak matches.
 */
export interface SemanticSearchOptions extends SearchOptions {
  /** Minimum cosine similarity (0..1) for a node to be returned. Default 0. */
  threshold?: number;
}

/**
 * Cosine similarity between two equal-length vectors.
 * Returns 0 when either vector has zero magnitude.
 *
 * Mirrors the Python `understand_core.embedding_search.cosine_similarity`.
 */
export function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let magA = 0;
  let magB = 0;
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    dot += a[i] * b[i];
    magA += a[i] * a[i];
    magB += b[i] * b[i];
  }
  magA = Math.sqrt(magA);
  magB = Math.sqrt(magB);
  if (magA === 0 || magB === 0) return 0;
  return dot / (magA * magB);
}

/**
 * Vector-embedding search engine using cosine similarity.
 *
 * Stores pre-computed per-node embeddings and ranks them against a query
 * embedding. Results share the lexical {@link SearchResult} shape so callers
 * can treat both engines uniformly (`score = 1 - similarity`; 0 = perfect).
 *
 * Mirrors the Python `understand_core.embedding_search.SemanticSearchEngine`.
 */
export class SemanticSearchEngine {
  private nodes: GraphNode[];
  private embeddings: Map<string, number[]>;

  constructor(nodes: GraphNode[], embeddings: Record<string, number[]>) {
    this.nodes = nodes;
    this.embeddings = new Map(Object.entries(embeddings));
  }

  hasEmbeddings(): boolean {
    return this.embeddings.size > 0;
  }

  addEmbedding(nodeId: string, embedding: number[]): void {
    this.embeddings.set(nodeId, embedding);
  }

  search(queryEmbedding: number[], options?: SemanticSearchOptions): SearchResult[] {
    const limit = options?.limit ?? 10;
    const threshold = options?.threshold ?? 0;
    const typeFilter = options?.types && options.types.length > 0 ? new Set(options.types) : null;

    const scored: SearchResult[] = [];
    for (const node of this.nodes) {
      if (typeFilter && !typeFilter.has(node.type)) continue;
      const embedding = this.embeddings.get(node.id);
      if (!embedding) continue;
      const similarity = cosineSimilarity(queryEmbedding, embedding);
      if (similarity >= threshold) {
        scored.push({ nodeId: node.id, score: 1 - similarity });
      }
    }

    scored.sort((a, b) => a.score - b.score);
    return scored.slice(0, limit);
  }

  updateNodes(nodes: GraphNode[]): void {
    this.nodes = nodes;
  }
}

File: understand-anything-plugin/packages/core/understand_core/embedding_search.py
"""Semantic (embedding) search over graph nodes.

Port of ``embedding-search.ts``. Cosine similarity is computed with ``numpy``.
Scores follow the same convention as :mod:`understand_core.search`: ``0`` is a
perfect match (similarity 1) and larger scores are worse.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Mapping, Sequence

import numpy as np

if TYPE_CHECKING:
    from understand_core.types import GraphNode

# Re-use the SearchResult shape from the lexical search module so callers can
# treat both engines uniformly.
from understand_core.search import SearchResult

__all__ = ["cosine_similarity", "SemanticSearchEngine", "SearchResult"]


def cosine_similarity(a: Sequence[float], b: Sequence[float]) -> float:
    """Cosine similarity between two vectors; 0 if either has zero magnitude."""
    va = np.asarray(a, dtype=np.float64)
    vb = np.asarray(b, dtype=np.float64)
    mag_a = float(np.linalg.norm(va))
    mag_b = float(np.linalg.norm(vb))
    if mag_a == 0.0 or mag_b == 0.0:
        return 0.0
    return float(np.dot(va, vb) / (mag_a * mag_b))


def _node_id(node: object) -> str:
    if isinstance(node, dict):
        return str(node.get("id", ""))
    return str(getattr(node, "id", ""))


def _node_type(node: object) -> str:
    if isinstance(node, dict):
        return str(node.get("type", ""))
    return str(getattr(node, "type", ""))


class SemanticSearchEngine:
    """Vector-embedding search engine using cosine similarity.

    Stores pre-computed embeddings for graph nodes and ranks them against a
    query embedding.
    """

    def __init__(
        self,
        nodes: Sequence[GraphNode],
        embeddings: Mapping[str, Sequence[float]],
    ) -> None:
        self._nodes: list[GraphNode] = list(nodes)
        self._embeddings: dict[str, list[float]] = {
            k: list(v) for k, v in embeddings.items()
        }

    def has_embeddings(self) -> bool:
        return len(self._embeddings) > 0

    def add_embedding(self, node_id: str, embedding: Sequence[float]) -> None:
        self._embeddings[node_id] = list(embedding)

    def search(
        self,
        query_embedding: Sequence[float],
        limit: int = 10,
        threshold: float = 0.0,
        types: Sequence[str] | None = None,
    ) -> list[SearchResult]:
        type_filter = set(types) if types else None

        scored: list[SearchResult] = []
        for node in self._nodes:
            if type_filter is not None and _node_type(node) not in type_filter:
                continue
            embedding = self._embeddings.get(_node_id(node))
            if embedding is None:
                continue
            similarity = cosine_similarity(query_embedding, embedding)
            if similarity >= threshold:
                scored.append(SearchResult(node_id=_node_id(node), score=1.0 - similarity))

        scored.sort(key=lambda r: r.score)
        return scored[:limit]

    def update_nodes(self, nodes: Sequence[GraphNode]) -> None:
        self._nodes = list(nodes)
